components/dashboard/ee/src/components/admin/user-view.tsx

```diff
@@ -150,6 +157,29 @@ export class UserView extends React.Component<UserViewProps, UserViewState> {
                 open={this.state.addingFeatureFlagOp === 'choose'}
                 onSelect={r => this.modifyWorkspaceFeatureFlags(r, true)}
             />
+            <Dialog open={this.state.deletingOp == 'confirmation'}>
+                <React.Fragment>
+                    <DialogTitle>Delete User: Are your sure?</DialogTitle>
+                    <DialogContent>
+                        <Typography variant="body1" style={{ width: "100%" }}>
+                            <div>Are you sure that you would like to delete user <strong>{this.state.user?.name}</strong>?</div>
+                            <div>
+                                Please type in the username <strong>{this.state.user?.name}</strong> to confirm: <br />
+                                <TextField value={this.state.deletingConfirmedUsername} onChange={e => this.setState({ deletingConfirmedUsername: (e.target as HTMLInputElement).value })} />
+                            </div>
+                        </Typography>
+                    </DialogContent>
+                    <DialogActions>
+                        <Button onClick={() => this.setState({ deletingOp: 'none' })} variant="outlined" color="primary">Cancel</Button>
```

Comment:
One should not use () => {} in handlers, it creates a new isntance each time and break react caches, i.e. it is always rerendered.

We need some react linters. I see such code in many our react components.

it should be like https://github.com/gitpod-io/gitpod/blob/afb017fa3b8c809967ca829e0cfff255376d542b/components/dashboard/src/components/user-settings.tsx#L28 the same instance of a function is used always

Reply:
Thanks a lot for this hint, @akosyakov! Good to know. :+1:

Reply:
Since I'm always keen to learn something new (who does not? :smile:), I had a quick Google research on this topic. It seems that this topic is quite controversial. My impression is that the opinion dominates that this is rather a premature optimization and not worth the more elaborate and cluttered code (especially for simple actions like a status updates).

Examples:

---

> It’s common to do nothing more than setState inside of event handlers for buttons, inputs, and other DOM components. This often makes an inline function the cleanest approach. Instead of bouncing around the file to find the event handlers, they’re colocated. The React community generally welcomes colocation.

> If you really believe that premature optimization is bad practice, then you won’t need proof that inline functions are fast, you need proof that they are slow.

_https://cdb.reacttraining.com/react-inline-functions-and-performance-bdff784f5578_

---


> A common rule of thumb is to measure performance of the app and only optimize if needed. Performance impact of inline function, often categorized under micro-optimizations, is always a tradeoff between code readability, performance gain, code organization, etc that must be thought out carefully on a case by case basis and pre-optimization should be avoided.
> 
> In this blog post, we observed that inline functions don't necessarily bring in a lot of performance cost. They are widely used because of ease of writing, reading and organizing inline functions, especially when inline function definitions are short and simple.

_https://www.velotio.com/engineering-blog/inline-function-performance-in-react-applications_

---

I tend to agree with this opinion. 

However, since you're far more experienced with React, I'm happy to change that if you're not convinced, @akosyakov. Otherwise, I would like to leave it as it is for cleaner code.

Reply:
> Since I'm always keen to learn something new (who does not? 😄)

👍 

> I tend to agree with this opinion.

The point of React to minify updates to DOM and boost performance with it, by binding new functions in render method we basically force React to always rerender this element in DOM instead. It can be not harmful for leaf elements like buttont, but misusage with composite elements can lead to constant rerendering of tree of child elements. I am not sure who is going to profile it and how since it will look like internal React code is slow. [Official docs](https://reactjs.org/docs/handling-events.html) recommend always avoid binding in render to eliminate such issues:

> The problem with this syntax is that a different callback is created each time the LoggingButton renders. In most cases, this is fine. However, if this callback is passed as a prop to lower components, those components might do an extra re-rendering. We generally recommend binding in the constructor or using the class fields syntax, to avoid this sort of performance problem.

> However, since you're far more experienced with React, I'm happy to change that if you're not convinced, @akosyakov. Otherwise, I would like to leave it as it is for cleaner code.

It is not up to me to decide :) In Theia we would discuss it with the team and agree on the convention like: https://github.com/eclipse-theia/theia/wiki/Coding-Guidelines#no-bind-fn-in-event-handlers After that noone can argue about it again to avoid sidetracking. It also helped new contributors since it prevents them introducing harmful code without having experience with React and so on. The bad side was that someone should remind on PRs, having eslint rules would be better, but noone had time to configure it.

In regards to readability mentioned articles seem to be concerned with readability with JS, but with TS we don't need bindings in the constructor and so on, we can assign function to a property and then there is no much difference:
- `{() => this.deleteUser()}` become `{this.deleteUser}`
- `protected async deleteUser() {` become `private deleteUser = async () => {` (btw it is better to use private, compiler will be detect and report when method is not used anymore)

That's said I don't want to sidetrack you in React debates on the PR if you feel that it is too much here then it is fine to merge like that. It can be addressed with introducing linting rule. Besides now you learn about this pitfall.

Reply:
Thanks for your insightful remarks! I changed the `deleteUser()` call as you suggested. I see, that there is no need for the inline function. I left the `setStatus(...)` calls as it is for now to avoid adding additional functions for each call. As you said, we can address this later when we have a consensus in the team and added linting rules for this.